Failing requirements update when scm_type != git

DanielDisisto opened the pull request:
##### SUMMARY
During a project update where SCM type is not `git`, and a `requirements.yml` exists within the checkout, a conditional check relating to `git` fails consistently and prevents the project update from ever completing.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
 - API

##### AWX VERSION
```
4.0.0
```


##### ADDITIONAL INFORMATION
```
The conditional check 'doesRequirementsExist.stat.exists and (scm_version is undefined or (git_result is defined and git_result['before'] == git_result['after']))' failed. The error was: error while evaluating conditional (doesRequirementsExist.stat.exists and (scm_version is undefined or (git_result is defined and git_result['before'] == git_result['after']))): 'dict object' has no attribute 'before'

The error appears to have been in '/var/lib/awx/venv/awx/lib/python3.6/site-packages/awx/playbooks/project_update.yml': line 142, column 9, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:


      - name: fetch galaxy roles from requirements.yml
        ^ here
```

DanielDisisto committed:
Failing requirements update when scm_type != git

The `git_result` variable is incorrectly checked as being `defined` vs. `not skipped`. This causes SVN (or any non-GIT) project update to fail consistently when `doesRequirementsExist.stat.exists` is true

Signed-off-by: DanielDisisto <daniel.disisto@didata.com.au>

AlanCoding commented:
I'm having a hard time reproducing this, I've tried subversion so far. What Ansible version is your server running, and what `scm_type` did you use in order to encounter this error? Could you share the rest of the project update playbook?

DanielDisisto commented:
Thanks for looking at this Alan. Using ansible version is 2.7.9 per the docker image `ansible/awx_task:4.0.0`. The only playbook relevant is `awx/playbooks/project_update.yml`, and the changes introduced in #3177. [Another commenter](https://github.com/ansible/awx/pull/3177#commitcomment-33047404) on this PR reported the same issue & findings, but never actually proposed a fix as far as I can tell.

This is 100% reproducible for me with -
* Project of `scm_type` of `svn` (or anything other than git)
* `AWX_ROLES_ENABLED=true`
* A `roles/requirements.yml` file inside the project / checkout

Variables set with `register` inside a block that is skipped get a value of `skipped` (which is indeed `defined`) -
```
(ansible) s:~/src/ansible$ cat test.yml
---
- hosts: localhost
  gather_facts: false
  tasks:

    - block:
      - git:
        register: git_result
      when: false

    - debug: var=git_result

(ansible) s:~/src/ansible$ ansible-playbook -i /dev/null test.yml
 [WARNING]: provided hosts list is empty, only localhost is available. Note
that the implicit localhost does not match 'all'


PLAY [localhost] ***************************************************************

TASK [git] *********************************************************************
skipping: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] => {
    "git_result": {
        "changed": false,
        "skip_reason": "Conditional result was False",
        "skipped": true
    }
}

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0
```

So the `git_result is defined` condition here:
https://github.com/ansible/awx/blob/4470e9ca26eb5500e71abcdc051956ec3c5a47d2/awx/playbooks/project_update.yml#L155
is _always_ going to be `true`, but doesn't have `before` & `after` attributes when skipped, which is why we get the exception `'dict object' has no attribute 'before'`.

AlanCoding commented:
I just cannot reproduce that. I installed Ansible 2.7.9 so that we'll be the same. I made a new repo with requirements file in the main branch to test as subversion, and ran it. Note that this led me to the bug https://github.com/ansible/awx/issues/4163, but this is not the same bug that you reported.

project update details:

```json
{
    "id": 8441,
    "type": "project_update",
    "url": "/api/v2/project_updates/8441/",
    "related": {
        "created_by": "/api/v2/users/1/",
        "unified_job_template": "/api/v2/projects/4153/",
        "stdout": "/api/v2/project_updates/8441/stdout/",
        "project": "/api/v2/projects/4153/",
        "cancel": "/api/v2/project_updates/8441/cancel/",
        "scm_inventory_updates": "/api/v2/project_updates/8441/scm_inventory_updates/",
        "notifications": "/api/v2/project_updates/8441/notifications/",
        "events": "/api/v2/project_updates/8441/events/"
    },
    "summary_fields": {
        "project": {
            "id": 4153,
            "name": "Roles checkout svn project ParkFault㼉",
            "description": "ConversationBreakfastPhaseSettingPaymentFixBlockSplitInterestResist�",
            "status": "successful",
            "scm_type": "svn"
        },
        "unified_job_template": {
            "id": 4153,
            "name": "Roles checkout svn project ParkFault㼉",
            "description": "ConversationBreakfastPhaseSettingPaymentFixBlockSplitInterestResist�",
            "unified_job_type": "project_update"
        },
        "instance_group": {
            "name": "tower",
            "id": 1
        },
        "created_by": {
            "id": 1,
            "username": "admin",
            "first_name": "",
            "last_name": ""
        },
        "user_capabilities": {
            "delete": true,
            "start": true
        }
    },
    "created": "2019-06-28T14:00:38.986074Z",
    "modified": "2019-06-28T14:00:39.283358Z",
    "name": "Roles checkout svn project ParkFault㼉",
    "description": "ConversationBreakfastPhaseSettingPaymentFixBlockSplitInterestResist�",
    "local_path": "_4153__roles_checkout_svn_project_parkfault",
    "scm_type": "svn",
    "scm_url": "https://github.com/AlanCoding/ansible-roles-example",
    "scm_branch": "",
    "scm_clean": false,
    "scm_delete_on_update": false,
    "credential": null,
    "timeout": 0,
    "unified_job_template": 4153,
    "launch_type": "manual",
    "status": "successful",
    "failed": false,
    "started": "2019-06-28T14:00:39.424832Z",
    "finished": "2019-06-28T14:00:48.490643Z",
    "elapsed": 9.066,
    "job_args": "[\"bwrap\", \"--unshare-pid\", \"--dev-bind\", \"/\", \"/\", \"--proc\", \"/proc\", \"--bind\", \"/tmp/ansible_runner_pi__8_6c7ob/tmpg7zl026q\", \"/awx_devel/awx/job_status\", \"--bind\", \"/tmp/ansible_runner_pi__8_6c7ob/tmpbv5_ipcb\", \"/etc/ssh\", \"--bind\", \"/tmp/ansible_runner_pi__8_6c7ob/tmpnq3hi2w1\", \"/projects\", \"--bind\", \"/tmp/ansible_runner_pi__8_6c7ob/tmpwqrm0nib\", \"/tmp\", \"--bind\", \"/tmp/ansible_runner_pi__8_6c7ob/tmpxfrir4gg\", \"/var/lib/awx\", \"--bind\", \"/tmp/ansible_runner_pi__8_6c7ob/tmp26z3hsm0\", \"/var/log\", \"--ro-bind\", \"/venv/ansible\", \"/venv/ansible\", \"--ro-bind\", \"/venv/awx\", \"/venv/awx\", \"--bind\", \"/awx_devel/awx/playbooks\", \"/awx_devel/awx/playbooks\", \"--bind\", \"/projects\", \"/projects\", \"--bind\", \"/tmp/awx_8441_uh4zp40t\", \"/tmp/awx_8441_uh4zp40t\", \"--chdir\", \"/awx_devel/awx/playbooks\", \"ansible-playbook\", \"-v\", \"-i\", \"/tmp/awx_8441_uh4zp40t/inventory/hosts\", \"-e\", \"@/tmp/awx_8441_uh4zp40t/env/extravars\", \"project_update.yml\"]",
    "job_cwd": "/awx_devel/awx/playbooks",
    "job_env": {
        "HOSTNAME": "awx",
        "MAKEFLAGS": "w",
        "RABBITMQ_USER": "guest",
        "OS": "Operating System: Docker for Mac",
        "LC_ALL": "en_US.UTF-8",
        "RABBITMQ_VHOST": "/",
        "SDB_HOST": "0.0.0.0",
        "MAKELEVEL": "2",
        "VIRTUAL_ENV": "/venv/ansible",
        "MFLAGS": "-w",
        "PATH": "/venv/ansible/bin:/venv/awx/bin:/venv/awx/bin:/usr/local/n/versions/node/10.15.0/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
        "RABBITMQ_PASS": "**********",
        "SUPERVISOR_GROUP_NAME": "tower-processes",
        "PWD": "/awx_devel",
        "LANG": "\"en-us\"",
        "PS1": "(awx) ",
        "SUPERVISOR_ENABLED": "1",
        "SHLVL": "2",
        "HOME": "/tmp",
        "LANGUAGE": "en_US:en",
        "AWX_GROUP_QUEUES": "tower",
        "SUPERVISOR_SERVER_URL": "unix:///tmp/supervisor.sock",
        "SUPERVISOR_PROCESS_NAME": "awx-dispatcher",
        "RABBITMQ_HOST": "rabbitmq",
        "CURRENT_UID": "501",
        "_": "/venv/awx/bin/python3",
        "DJANGO_SETTINGS_MODULE": "awx.settings.development",
        "DJANGO_LIVE_TEST_SERVER_ADDRESS": "localhost:9013-9199",
        "SDB_NOTIFY_HOST": "docker.for.mac.host.internal",
        "TZ": "UTC",
        "ANSIBLE_FORCE_COLOR": "True",
        "ANSIBLE_HOST_KEY_CHECKING": "False",
        "ANSIBLE_INVENTORY_UNPARSED_FAILED": "True",
        "ANSIBLE_PARAMIKO_RECORD_HOST_KEYS": "False",
        "ANSIBLE_VENV_PATH": "/venv/ansible",
        "PROOT_TMP_DIR": "/tmp",
        "AWX_PRIVATE_DATA_DIR": "/tmp/awx_8441_uh4zp40t",
        "PYTHONPATH": "/venv/ansible/lib/python2.7/site-packages:",
        "ANSIBLE_RETRY_FILES_ENABLED": "False",
        "ANSIBLE_ASK_PASS": "False",
        "ANSIBLE_BECOME_ASK_PASS": "False",
        "DISPLAY": "",
        "TMP": "/tmp",
        "PROJECT_UPDATE_ID": "8441",
        "ANSIBLE_CALLBACK_PLUGINS": "/awx_devel/awx/plugins/callback",
        "ANSIBLE_STDOUT_CALLBACK": "awx_display",
        "AWX_ISOLATED_DATA_DIR": "/tmp/awx_8441_uh4zp40t/artifacts/8441"
    },
    "job_explanation": "",
    "execution_node": "awx",
    "result_traceback": "",
    "event_processing_finished": true,
    "project": 4153,
    "job_type": "check",
    "host_status_counts": {
        "changed": 1
    },
    "playbook_counts": {
        "play_count": 2,
        "task_count": 18
    }
}
```

The playbook output:

```
Using /etc/ansible/ansible.cfg as config file
/tmp/awx_8441_uh4zp40t/inventory/hosts did not meet host_list requirements, check plugin documentation if this is unexpected
/tmp/awx_8441_uh4zp40t/inventory/hosts did not meet script requirements, check plugin documentation if this is unexpected


PLAY [all] *********************************************************************

TASK [delete project directory before update] **********************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [check repo using git] ****************************************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [update project using git] ************************************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Set the git repository version] ******************************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [update project using hg] *************************************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Set the hg repository version] *******************************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [parse hg version string properly] ****************************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [update project using svn] ************************************************
ok: [localhost, -> localhost] => {"after": ["Revision: 2", "URL: https://github.com/AlanCoding/ansible-roles-example"], "before": ["Revision: 2", "URL: https://github.com/AlanCoding/ansible-roles-example"], "changed": false}

TASK [Set the svn repository version] ******************************************
ok: [localhost, -> localhost] => {"ansible_facts": {"scm_version": ["Revision: 2", "URL: https://github.com/AlanCoding/ansible-roles-example"]}, "changed": false}

TASK [parse subversion version string properly] ********************************
ok: [localhost, -> localhost] => {"ansible_facts": {"scm_version": "2"}, "changed": false}

TASK [Ensure the project directory is present] *********************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Fetch Insights Playbook(s)] **********************************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Save Insights Version] ***************************************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Repository Version] ******************************************************
ok: [localhost,] => {
    "msg": "Repository Version 2"
}

TASK [Write Repository Version] ************************************************
changed: [localhost, -> localhost] => {"changed": true, "checksum": "da4b9237bacccdf19c0760cab7aec4a8359010b0", "dest": "/projects/tmpbmla56qd", "gid": 0, "group": "root", "md5sum": "c81e728d9d4c2f636f067f89cc14862c", "mode": "0600", "owner": "awx", "size": 1, "src": "/tmp/.ansible/tmp/ansible-tmp-1561730447.1-87308454354967/source", "state": "file", "uid": 501}

PLAY [all] *********************************************************************

TASK [detect requirements.yml] *************************************************
ok: [localhost, -> localhost] => {"changed": false, "stat": {"exists": false}}

TASK [fetch galaxy roles from requirements.yml] ********************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [fetch galaxy roles from requirements.yml (forced update)] ****************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

PLAY RECAP *********************************************************************
localhost,                 : ok=6    changed=1    unreachable=0    failed=0   
```

Nothing jumps out to me about what I might be doing differently from you. I'm kind of stumped.

DanielDisisto commented:
Ok I see the reason why that didn't reproduce the issue - in the `detect requirements.yml` task, your run doesn't seem to find the file, so first check on line 155 `doesRequirementsExist.stat.exists` is false & the rest is not evaluated. When I run it looks like this (from that task onwards) -

```
[snip]

TASK [detect requirements.yml] *************************************************
ok: [localhost,] => {"changed": false, "stat": {"atime": 1561678685.498315, "attr_flags": "", "attributes": [], "block_size": 4096, "blocks": 0, "charset": "unknown", "ctime": 1547509514.4089012, "dev": 65028, "device_type": 0, "executable": false, "exists": true, "gid": 1350800513, "inode": 456670, "isblk": false, "ischr": false, "isdir": false, "isfifo": false, "isgid": false, "islnk": true, "isreg": false, "issock": false, "isuid": false, "lnk_source": "/var/lib/awx/projects/_6__gsoa_svn/requirements.yml", "lnk_target": "../requirements.yml", "mimetype": "unknown", "mode": "0777", "mtime": 1544059510.4352922, "nlink": 1, "path": "/var/lib/awx/projects/_6__gsoa_svn/roles/requirements.yml", "pw_name": "root", "readable": true, "rgrp": true, "roth": true, "rusr": true, "size": 19, "uid": 0, "version": null, "wgrp": true, "woth": true, "writeable": true, "wusr": true, "xgrp": true, "xoth": true, "xusr": true}}

TASK [fetch galaxy roles from requirements.yml] ********************************
fatal: [localhost,]: FAILED! => {"msg": "The conditional check 'doesRequirementsExist.stat.exists and (scm_version is undefined or (git_result is defined and git_result['before'] == git_result['after']))' failed. The error was: error while evaluating conditional (doesRequirementsExist.stat.exists and (scm_version is undefined or (git_result is defined and git_result['before'] == git_result['after']))): 'dict object' has no attribute 'before'\\n\\nThe error appears to have been in '/var/lib/awx/venv/awx/lib/python3.6/site-packages/awx/playbooks/project_update.yml': line 142, column 9, but may\\nbe elsewhere in the file depending on the exact syntax problem.\\n\\nThe offending line appears to be:\\n\\n\\n      - name: fetch galaxy roles from requirements.yml\\n        ^ here\\n"}

PLAY RECAP *********************************************************************
localhost,                 : ok=6    changed=2    unreachable=0    failed=1
```

Where that requirements file contains galaxy roles required in our project, i.e.
```
s:.../projects/_6__gsoa_svn$ cat roles/requirements.yml
- src: http://<internal fqdn>/store/galaxy/cchurch.admin-users/files.tar.gz
  name: cchurch.admin-users
...
```

AlanCoding commented:
Are you using mercurial? I see how that could be consistent with all the other information here.

AlanCoding commented:
well I went ahead and created an example on my own that does this with mercurial

https://bitbucket.org/AlanCoding/ansible-roles-example-hg

I have found that this does reproduce the issue.

DanielDisisto commented:
Sorry for not providing more direct reproduction instructions from the start. I've just discovered that you can also do subversion checkouts directly from a github repo, so one such as https://github.com/AgentCormac/ProjectX-Ansible can also be used to reproduce the error. If you do want to see it occur with subversion too, a project setup such as this will reliably do so -
```
scm_type: svn
scm_url: https://github.com/AgentCormac/ProjectX-Ansible/trunk
```
```
Using /etc/ansible/ansible.cfg as config file
/tmp/awx_3201_u1i_48wa/inventory/hosts did not meet host_list requirements, check plugin documentation if this is unexpected
/tmp/awx_3201_u1i_48wa/inventory/hosts did not meet script requirements, check plugin documentation if this is unexpected


PLAY [all] *********************************************************************

TASK [delete project directory before update] **********************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [check repo using git] ****************************************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [update project using git] ************************************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Set the git repository version] ******************************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [update project using hg] *************************************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Set the hg repository version] *******************************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [parse hg version string properly] ****************************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [update project using svn] ************************************************
changed: [localhost,] => {"after": ["Revision: 38", "URL: https://github.com/AgentCormac/ProjectX-Ansible/trunk"], "before": null, "changed": true}

TASK [Set the svn repository version] ******************************************
ok: [localhost,] => {"ansible_facts": {"scm_version": ["Revision: 38", "URL: https://github.com/AgentCormac/ProjectX-Ansible/trunk"]}, "changed": false}

TASK [parse subversion version string properly] ********************************
ok: [localhost,] => {"ansible_facts": {"scm_version": "38"}, "changed": false}

TASK [Ensure the project directory is present] *********************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Fetch Insights Playbook(s)] **********************************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Save Insights Version] ***************************************************
skipping: [localhost,] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Repository Version] ******************************************************
ok: [localhost,] => {
    "msg": "Repository Version 38"
}

TASK [Write Repository Version] ************************************************
changed: [localhost,] => {"changed": true, "checksum": "5b384ce32d8cdef02bc3a139d4cac0a22bb029e8", "dest": "/var/lib/awx/projects/tmp0b4x21yr", "gid": 0, "group": "root", "md5sum": "a5771bce93e200c36f7cd9dfd0e5deaa", "mode": "0600", "owner": "root", "size": 2, "src": "/root/.ansible/tmp/ansible-tmp-1562032274.78-271019842522873/source", "state": "file", "uid": 0}

PLAY [all] *********************************************************************

TASK [detect requirements.yml] *************************************************
ok: [localhost,] => {"changed": false, "stat": {"atime": 1562032272.4939668, "attr_flags": "", "attributes": [], "block_size": 4096, "blocks": 8, "charset": "unknown", "checksum": "9ce503c998683f65cce17ff4260943868e5957cb", "ctime": 1562032272.4939668, "dev": 65028, "device_type": 0, "executable": false, "exists": true, "gid": 0, "gr_name": "root", "inode": 1919574, "isblk": false, "ischr": false, "isdir": false, "isfifo": false, "isgid": false, "islnk": false, "isreg": true, "issock": false, "isuid": false, "mimetype": "unknown", "mode": "0644", "mtime": 1562032272.4939668, "nlink": 1, "path": "/var/lib/awx/projects/_52__svn_project_test/roles/requirements.yml", "pw_name": "root", "readable": true, "rgrp": true, "roth": true, "rusr": true, "size": 106, "uid": 0, "version": null, "wgrp": false, "woth": false, "writeable": true, "wusr": true, "xgrp": false, "xoth": false, "xusr": false}}

TASK [fetch galaxy roles from requirements.yml] ********************************
fatal: [localhost,]: FAILED! => {"msg": "The conditional check 'doesRequirementsExist.stat.exists and (scm_version is undefined or (git_result is defined and git_result['before'] == git_result['after']))' failed. The error was: error while evaluating conditional (doesRequirementsExist.stat.exists and (scm_version is undefined or (git_result is defined and git_result['before'] == git_result['after']))): 'dict object' has no attribute 'before'\\n\\nThe error appears to have been in '/var/lib/awx/venv/awx/lib/python3.6/site-packages/awx/playbooks/project_update.yml': line 142, column 9, but may\\nbe elsewhere in the file depending on the exact syntax problem.\\n\\nThe offending line appears to be:\\n\\n\\n      - name: fetch galaxy roles from requirements.yml\\n        ^ here\\n"}

PLAY RECAP *********************************************************************
localhost,                 : ok=6    changed=2    unreachable=0    failed=1   
```

AlanCoding commented:
Could you do the instructions to do the sign-off on your commit?

https://github.com/ansible/awx/blob/devel/DCO_1_1.md

Thanks

DanielDisisto commented:
Sure, have added the sign-off to the commit.
